### PR TITLE
Allow PoolManager and ConnectionPool to be used as a context manager.

### LINKED ISSUE #514: Add Context Managers
The classes like ``PoolManager`` and ``ConnectionPool`` and such should act as context managers to make it easier to clean up the resources used by them.

<bountysource-plugin>

---
Want to back this issue? **[Place a bounty on it!](https://www.bountysource.com/issues/6534710-add-context-managers?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sornars**
Fixes #514 

I'm not sure if any behaviour beyond closing/clearing the pools is expected from a context manager, let me know if I've missed the mark.

### CODE REVIEW

**Comment by Lukasa** [comment]
This is a good idea!

I think implementing on the base class is a good idea.

**Comment by sigmavirus24** [comment]
Would you mind commenting that "return False" causes the original exception to be automatically re-raised if one occurs in the context? I always confuse False with True here since True indicates that the error should be swallowed instead. This will probably help others without having to re-read the PEP (like I just did).

**Comment by shazow** [comment]
I love it!

Would you like to add a usage example to the generated docs? (Happy to merge now if you'd like to do that separately.)

**Comment by sornars** [comment]
I've gone ahead and updated the docs, let me know if you have any suggestions. 

I'm a bit confused by the failing builds, the listed errors in the failed builds don't seem to have anything to do with the changes I've made. Should I be concerned?

**Comment by shazow** [comment]
Re-ran the tests. Unfortunately they're kinda flakey right now. :( (Would really appreciate any help, but that's unrelated to this PR).

File: docs/pools.rst [hunk 1]
```diff
@@ -33,7 +33,21 @@ If you need to make requests to the same host repeatedly, then you should use a
 By default, the pool will cache just one connection. If you're planning on using
 such a pool in a multithreaded environment, you should set the ``maxsize`` of
 the pool to a higher number, such as the number of threads. You can also control
-many other variables like timeout, blocking, and default headers.
+many other variables like timeout, blocking, and default headers. Additionally,
```
Comment (shazow) [inline docs/pools.rst]: Would give it more visibility if you broke it out into a new paragraph.

File: docs/pools.rst [hunk 2]
```diff
@@ -33,7 +33,21 @@ If you need to make requests to the same host repeatedly, then you should use a
...
 the pool to a higher number, such as the number of threads. You can also control
-many other variables like timeout, blocking, and default headers.
+many other variables like timeout, blocking, and default headers. Additionally, 
+you can use a ConnectionPool as a context manager so the pool will be automatically
+cleared after usage.
+
+.. doctest ::
+
+    >>> from urllib3 import HTTPConnectionPool
+    >>> with  HTTPConnectionPool('ajax.googleapis.com', maxsize=1) as pool:
```
Comment (shazow) [inline docs/pools.rst]: Superfluous whitespace after `with`.

**Comment by shazow** [comment]
Thank you supermuch. :)

**Pull request merged**